From a3d844a14a599496946418bd4af0ed01d3200309 Mon Sep 17 00:00:00 2001
From: Giuseppe CAVALLARO <peppe.cavallaro@st.com>
Date: Tue, 7 Jul 2009 16:26:38 +0200
Subject: [PATCH] phy: update the SMSC PHY driver

This patch updates the SMSC PHY device driver to
the latest version from net-2.6.git.
It also adds the SMSC LAN 8710 support.

Signed-off-by: Giuseppe Cavallaro <peppe.cavallaro@st.com>
Signed-off-by: Stuart Menefy <stuart.menefy@st.com>
---
 drivers/net/phy/smsc.c |  245 +++++++++++++++++++++++++++++++++---------------
 1 files changed, 171 insertions(+), 74 deletions(-)

diff --git a/drivers/net/phy/smsc.c b/drivers/net/phy/smsc.c
index 87ac3d7..03052ae 100644
--- a/drivers/net/phy/smsc.c
+++ b/drivers/net/phy/smsc.c
@@ -12,9 +12,8 @@
  * Free Software Foundation;  either version 2 of the  License, or (at your
  * option) any later version.
  *
- * Changelog:
- *	March 2008
- *	- Added LAN8700 PHY (Giuseppe Cavallaro <peppe.cavallaro@st.com>)
+ * Support added for SMSC LAN8187 and LAN8700 by steve.glendinning@smsc.com
+ *
  */
 
 #include <linux/kernel.h>
@@ -24,110 +23,208 @@
 #include <linux/phy.h>
 #include <linux/netdevice.h>
 
-#define MII_SMSC_ISF 29 /* Interrupt Source Flags */
-#define MII_SMSC_IM  30 /* Interrupt Mask */
-
-#define MII_SMSC_ISF_INT1 (1<<1) /* Auto-Negotiation Page Received */
-#define MII_SMSC_ISF_INT2 (1<<2) /* Parallel Detection Fault */
-#define MII_SMSC_ISF_INT3 (1<<3) /* Auto-Negotiation LP Ack */
-#define MII_SMSC_ISF_INT4 (1<<4) /* Link Down */
-#define MII_SMSC_ISF_INT5 (1<<5) /* Remote Fault Detected */
-#define MII_SMSC_ISF_INT6 (1<<6) /* Auto-Negotiation complete */
-#define MII_SMSC_ISF_INT7 (1<<7) /* ENERGYON */
-
-#define MII_SMSC_ISF_INT_ALL (0x0e)
-
-#define MII_SMSC_ISF_INT_PHYLIB_EVENTS \
-	(MII_SMSC_ISF_INT6 | MII_SMSC_ISF_INT4)
-
-/* LAN8700 phy identifier values */
-#define LAN8700_PHY_ID		0x0007c0c0
-#define LAN8700_PHY_LO_ID_REVA 	0xc0c1
-#define LAN8700_PHY_LO_ID_REVB 	0xc0c2
-#define LAN8700_PHY_LO_ID_REVC 	0xc0c3
-#define LAN8700_PHY_LO_ID_REVD 	0xc0c4
-#define LAN8700_PHY_LO_ID_REVE 	0xc0c5
-#define LAN8700_PHY_LO_ID_REVF 	0xc0c6
-#define LAN8700_PHY_LO_ID_REVG 	0xc0c7
-
-static int smsc_config_intr(struct phy_device *phydev)
+#define MII_LAN83C185_ISF 29	/* Interrupt Source Flags */
+#define MII_LAN83C185_IM  30	/* Interrupt Mask */
+
+#define MII_LAN83C185_ISF_INT1 (1<<1)	/* Auto-Negotiation Page Received */
+#define MII_LAN83C185_ISF_INT2 (1<<2)	/* Parallel Detection Fault */
+#define MII_LAN83C185_ISF_INT3 (1<<3)	/* Auto-Negotiation LP Ack */
+#define MII_LAN83C185_ISF_INT4 (1<<4)	/* Link Down */
+#define MII_LAN83C185_ISF_INT5 (1<<5)	/* Remote Fault Detected */
+#define MII_LAN83C185_ISF_INT6 (1<<6)	/* Auto-Negotiation complete */
+#define MII_LAN83C185_ISF_INT7 (1<<7)	/* ENERGYON */
+
+#define MII_LAN83C185_ISF_INT_ALL (0x0e)
+
+#define MII_LAN83C185_ISF_INT_PHYLIB_EVENTS \
+	(MII_LAN83C185_ISF_INT6 | MII_LAN83C185_ISF_INT4)
+
+static int smsc_phy_config_intr(struct phy_device *phydev)
 {
-	int rc = phy_write (phydev, MII_SMSC_IM,
-			((PHY_INTERRUPT_ENABLED == phydev->interrupts)
-			? MII_SMSC_ISF_INT_PHYLIB_EVENTS
-			: 0));
+	int rc = phy_write(phydev, MII_LAN83C185_IM,
+			   ((PHY_INTERRUPT_ENABLED == phydev->interrupts)
+			    ? MII_LAN83C185_ISF_INT_PHYLIB_EVENTS : 0));
 
 	return rc < 0 ? rc : 0;
 }
 
-static int smsc_ack_interrupt(struct phy_device *phydev)
+static int smsc_phy_ack_interrupt(struct phy_device *phydev)
 {
-	int rc = phy_read (phydev, MII_SMSC_ISF);
+	int rc = phy_read(phydev, MII_LAN83C185_ISF);
 
 	return rc < 0 ? rc : 0;
 }
 
-static int smsc_config_init(struct phy_device *phydev)
+static int smsc_phy_config_init(struct phy_device *phydev)
 {
-	return smsc_ack_interrupt (phydev);
+	return smsc_phy_ack_interrupt(phydev);
 }
 
-
 static struct phy_driver lan83c185_driver = {
-	.phy_id		= 0x0007c0a0, /* OUI=0x00800f, Model#=0x0a */
-	.phy_id_mask	= 0xfffffff0,
-	.name		= "SMSC LAN83C185",
+	.phy_id = 0x0007c0a0,	/* OUI=0x00800f, Model#=0x0a */
+	.phy_id_mask = 0xfffffff0,
+	.name = "SMSC LAN83C185",
 
-	.features	= (PHY_BASIC_FEATURES | SUPPORTED_Pause
-				| SUPPORTED_Asym_Pause),
-	.flags		= PHY_HAS_INTERRUPT | PHY_HAS_MAGICANEG,
+	.features = (PHY_BASIC_FEATURES | SUPPORTED_Pause
+		     | SUPPORTED_Asym_Pause),
+	.flags = PHY_HAS_INTERRUPT | PHY_HAS_MAGICANEG,
 
 	/* basic functions */
-	.config_aneg	= genphy_config_aneg,
-	.read_status	= genphy_read_status,
-	.config_init	= smsc_config_init,
+	.config_aneg = genphy_config_aneg,
+	.read_status = genphy_read_status,
+	.config_init = smsc_phy_config_init,
 
 	/* IRQ related */
-	.ack_interrupt	= smsc_ack_interrupt,
-	.config_intr	= smsc_config_intr,
+	.ack_interrupt = smsc_phy_ack_interrupt,
+	.config_intr = smsc_phy_config_intr,
 
-	.driver		= { .owner = THIS_MODULE, }
+	.suspend = genphy_suspend,
+	.resume = genphy_resume,
+
+	.driver = {.owner = THIS_MODULE,}
+};
+
+static struct phy_driver lan8187_driver = {
+	.phy_id = 0x0007c0b0,	/* OUI=0x00800f, Model#=0x0b */
+	.phy_id_mask = 0xfffffff0,
+	.name = "SMSC LAN8187",
+
+	.features = (PHY_BASIC_FEATURES | SUPPORTED_Pause
+		     | SUPPORTED_Asym_Pause),
+	.flags = PHY_HAS_INTERRUPT | PHY_HAS_MAGICANEG,
+
+	/* basic functions */
+	.config_aneg = genphy_config_aneg,
+	.read_status = genphy_read_status,
+	.config_init = smsc_phy_config_init,
+
+	/* IRQ related */
+	.ack_interrupt = smsc_phy_ack_interrupt,
+	.config_intr = smsc_phy_config_intr,
+
+	.suspend = genphy_suspend,
+	.resume = genphy_resume,
+
+	.driver = {.owner = THIS_MODULE,}
 };
 
 static struct phy_driver lan8700_driver = {
-        .phy_id         = LAN8700_PHY_ID,
-        .phy_id_mask    = 0xfffffff0,
-        .name           = "SMSC LAN8700",
-        .features       = PHY_BASIC_FEATURES | SUPPORTED_Pause
-				| SUPPORTED_Asym_Pause,
-        .flags          = PHY_HAS_INTERRUPT,
-	.config_init    = smsc_config_init,
-        .config_aneg    = genphy_config_aneg,
-        .read_status    = genphy_read_status,
-        .ack_interrupt  = smsc_ack_interrupt,
-        .config_intr    = smsc_config_intr,
-	.suspend 	= genphy_suspend,
-	.resume  	= genphy_resume,
-	.driver         = { .owner = THIS_MODULE, }
+	.phy_id = 0x0007c0c0,	/* OUI=0x00800f, Model#=0x0c */
+	.phy_id_mask = 0xfffffff0,
+	.name = "SMSC LAN8700",
+
+	.features = (PHY_BASIC_FEATURES | SUPPORTED_Pause
+		     | SUPPORTED_Asym_Pause),
+	.flags = PHY_HAS_INTERRUPT | PHY_HAS_MAGICANEG,
+
+	/* basic functions */
+	.config_aneg = genphy_config_aneg,
+	.read_status = genphy_read_status,
+	.config_init = smsc_phy_config_init,
+
+	/* IRQ related */
+	.ack_interrupt = smsc_phy_ack_interrupt,
+	.config_intr = smsc_phy_config_intr,
+
+	.suspend = genphy_suspend,
+	.resume = genphy_resume,
+
+	.driver = {.owner = THIS_MODULE,}
 };
 
+static struct phy_driver lan911x_int_driver = {
+	.phy_id = 0x0007c0d0,	/* OUI=0x00800f, Model#=0x0d */
+	.phy_id_mask = 0xfffffff0,
+	.name = "SMSC LAN911x Internal PHY",
 
-static int __init smsc_init(void)
-{
-	int retval;
+	.features = (PHY_BASIC_FEATURES | SUPPORTED_Pause
+		     | SUPPORTED_Asym_Pause),
+	.flags = PHY_HAS_INTERRUPT | PHY_HAS_MAGICANEG,
+
+	/* basic functions */
+	.config_aneg = genphy_config_aneg,
+	.read_status = genphy_read_status,
+	.config_init = smsc_phy_config_init,
+
+	/* IRQ related */
+	.ack_interrupt = smsc_phy_ack_interrupt,
+	.config_intr = smsc_phy_config_intr,
+
+	.suspend = genphy_suspend,
+	.resume = genphy_resume,
+
+	.driver = {.owner = THIS_MODULE,}
+};
+
+static struct phy_driver lan8710_driver = {
+	.phy_id = 0x0007c0f0,	/* OUI=0x00800f, Model#=0x0f */
+	.phy_id_mask = 0xfffffff0,
+	.name = "SMSC LAN8710/LAN8720",
+
+	.features = (PHY_BASIC_FEATURES | SUPPORTED_Pause
+		     | SUPPORTED_Asym_Pause),
+	.flags = PHY_HAS_INTERRUPT | PHY_HAS_MAGICANEG,
+
+	/* basic functions */
+	.config_aneg = genphy_config_aneg,
+	.read_status = genphy_read_status,
+	.config_init = smsc_phy_config_init,
 
-	retval = phy_driver_register (&lan83c185_driver);
+	/* IRQ related */
+	.ack_interrupt = smsc_phy_ack_interrupt,
+	.config_intr = smsc_phy_config_intr,
+
+	.suspend = genphy_suspend,
+	.resume = genphy_resume,
 
-	if (retval < 0)
-		return retval;
+	.driver = {.owner = THIS_MODULE,}
+};
 
-	return phy_driver_register(&lan8700_driver);
+static int __init smsc_init(void)
+{
+	int ret;
+
+	ret = phy_driver_register(&lan83c185_driver);
+	if (ret)
+		goto err1;
+
+	ret = phy_driver_register(&lan8187_driver);
+	if (ret)
+		goto err2;
+
+	ret = phy_driver_register(&lan8700_driver);
+	if (ret)
+		goto err3;
+
+	ret = phy_driver_register(&lan911x_int_driver);
+	if (ret)
+		goto err4;
+
+	ret = phy_driver_register(&lan8710_driver);
+	if (ret)
+		goto err5;
+
+	return 0;
+
+err5:
+	phy_driver_unregister(&lan911x_int_driver);
+err4:
+	phy_driver_unregister(&lan8700_driver);
+err3:
+	phy_driver_unregister(&lan8187_driver);
+err2:
+	phy_driver_unregister(&lan83c185_driver);
+err1:
+	return ret;
 }
 
 static void __exit smsc_exit(void)
 {
-	phy_driver_unregister (&lan83c185_driver);
-	phy_driver_unregister (&lan8700_driver);
+	phy_driver_unregister(&lan8710_driver);
+	phy_driver_unregister(&lan911x_int_driver);
+	phy_driver_unregister(&lan8700_driver);
+	phy_driver_unregister(&lan8187_driver);
+	phy_driver_unregister(&lan83c185_driver);
 }
 
 MODULE_DESCRIPTION("SMSC PHY driver");
-- 
1.6.0.6

